### ISSUE TITLE
[rule] `jsx-no-trailing-semicolon`: add a rule to disallow trailing semicolons in JSX

### ISSUE THREAD

**Issue by nstepien** [Rel1cx/eslint-react#1685]
### Problem Description

It can happen that trailing semicolons are unintentionally left in JSX, for example when wrapping a one-liner in another jsx element/fragment:
```jsx
// before
return <div />;

// after
return (
  <div>
    <div />;
  </div>
);
```
this will then render a text node with `;` in the DOM, which is not intended in 99.99% of cases.

![Recording 2023-11-27 at 23 16 29](https://github.com/biomejs/biome/assets/567105/5993bfce-4ea2-4119-8d46-c48d25bccc35)

Biome has a rule for this:
https://biomejs.dev/linter/rules/no-suspicious-semicolon-in-jsx/
The original rule request in Biome:
https://github.com/biomejs/biome/issues/927

### Alternative Solutions

🤷‍♂️

## Rule Details

When refactoring JSX, trailing semicolons may be kept unintentionally immediately after JSX elements or fragments. This leads to `;` semicolons being unexpectedly rendered as text nodes in the DOM.

### Invalid

```tsx
// ❌ Bad: a trailing semicolon will be rendered in the DOM
return (
  <div>
    <div />;
  </div>
);

// ❌ Bad: a trailing semicolon will be rendered in the DOM
return (
  <div>
    <Component>
      <div />
    </Component>;
  </div>
);
```

### Valid

```tsx
// ✅ Good: no issues
return (
  <div>
    <div />
  </div>
);

// ✅ Good: rendering a semicolon on its own line is more explicit, it may be intentional, it is easier to catch during code reviews
return (
  <div>
    <div />
    ;
  </div>
);

// ✅ Good: rendering a semicolon in a JSX expression is more explicit, it may be intentional, it is easier to catch during code reviews
return (
  <div>
    <div />{';'}
  </div>
);

// ✅ Good: rendering a semicolon using an HTML entity is more explicit, it may be intentional, it is easier to catch during code reviews
return (
  <div>
    <div />&#59;
  </div>
);

// ✅ Good: semicolon in JSX within an element, between elements, after text, or as part of an HTML entity are allowed
return (
  <div>
    <span>;</span>
    text;
    <span />;<span />
    &amp;
  </div>
);

// ✅ Good: this semicolon is outside of the JSX tree and will not be rendered in the DOM
return <div />;

// ✅ Good: this semicolon is outside of the JSX tree and will not be rendered in the DOM
return <div>
  <div />
</div>;
```

### Why is this pattern problematic?

Unintentionally rendering semicolons in the DOM will lead to unexpected layout issues, and may reveal a potential oversight.

## Resources

- [Rule Source](https://github.com/Rel1cx/eslint-react/tree/main/packages/plugins/eslint-plugin-react-jsx/src/rules/no-trailing-semicolon)

### Evaluation Checklist

- [x] I have had problems with the pattern I want to forbid
- [x] I could not find a way to solve the problem by changing the API of the problematic code or introducing a new API
- [x] I have thought very hard about what the corner cases could be and what kinds of patterns this would forbid that are actually okay, and they are acceptable
- [x] I think the rule explains well enough how to solve the issue to make sure beginners are not blocked by it
- [x] I have discussed this rule with team members, and they all find it valuable

**Comment by Rel1cx** [comment]
This rule can be added to [packages/plugins/eslint-plugin-react-jsx/src/rules/no-trailing-semicolon](https://github.com/Rel1cx/eslint-react/blob/main/packages/plugins/eslint-plugin-react-jsx/src/rules).

  The rule implementation should be similar to:
  https://github.com/Rel1cx/eslint-react/blob/main/packages/plugins/eslint-plugin-react-jsx/src/rules/no-comment-textnodes/no-comment-textnodes.ts
